✨Create new project without internet

omegaStag opened the pull request:
#### Summary:
- Added 'download-ok' switch to ``new-project`` to decide if remote templates can be used in creating projects
- Put some lines from ``new-project`` into new helper function ``_create_project`` for this purpose (and to make click happy)
- Due to a generic Exception being raised when there is no internet connection (instead of an actual ConnectionError for some reason), added a nested try and except statement to attempt a new project with local templates if remote templates cannot be reached, raises another exception if that fails
- Fixed ``resolved_templates`` so that it actually takes into account whether 'download-ok' is true or false before downloading from remote
- "Can't GitHub" error could be placed before the output telling user which templates are used, but I'm not sure how to place it as I don't want it to trigger before a potential Exception due to something that is NOT a lack of internet
- Removed ``allow_online: bool = True`` from top of ``resolve_templates`` to make above features work
- Removed ``allow_online=download_ok`` from ``template = self.resolve_template(identifier=identifier, allow_online=download_ok, **kwargs)`` because `resolve_templates` now takes 'download_ok' into account seperately


#### Motivation:
Let users make new project from local templates without internet

##### References (optional):
Closes #91 

#### Test Plan:
1. **(With internet)** Purge templates, download 2nd most recent versions of okapilib and kernel, then run ``pros c n . --no-download`` to demonstrate that the most recent remote templates are not applied, and the local templates are used instead
![image](https://user-images.githubusercontent.com/57331513/139360629-ff7f30ef-6670-4cdc-8809-9bb4c6422831.png)


2. **(Without internet)** Create a new project with ``pros c n .``. There will be an error that alerts the user that the CLI can't connect to GitHub, and it creates a project with the local templates you have instead.
![image](https://user-images.githubusercontent.com/57331513/139360942-806501f6-60ee-4efd-bd4f-fbc8dcae64f6.png)


3. **(Without internet)** ``pros c n . --download``. Even if --download is specified, if there is no internet it creates from local templates without problem.
![image](https://user-images.githubusercontent.com/57331513/139363182-b9bcf776-8348-4698-b573-f6dafae322e9.png)


4. **(With internet)** ``pros c n .``. Newest versions of templates from remote are automatically downloaded and applied as usual.
![image](https://user-images.githubusercontent.com/57331513/139361488-889354c1-822a-440f-a0f4-fbededcfa69e.png)


5. **(Without internet)** ``pros c n .`` creates a new project with the most recent local templates.
![image](https://user-images.githubusercontent.com/57331513/139361450-257c2883-b9b9-48cb-9c9e-90ce2f719d38.png)

omegaStag committed:
Add allow-online switch to new-project

new-project now has an --allow-online/--no-online switch that allows for creating a project with or without remote templates.
also, made a helper _create_project function that new_project calls
FIXME: ConnectionErrors are not being caught (to make projects with local templates if no internet), unpoggers!

omegaStag committed:
Update conductor.py

omegaStag committed:
Added download-ok switch

Will elaborate in PR

WillXuCodes reviewed (dismissed):
Unless anyone else has any other comments, this seems to work and be good based on the test cases.

edjubuh commented:
I'd rather just fix `resolve_templates` to always be OK if a remote depot is unavailable. 

e.g. try block around:
https://github.com/purduesigbots/pros-cli/blob/1d9f6bae72b7e44ca4329d337b3921b7c41ae6df/pros/conductor/conductor.py#L119-L124

omegaStag committed:
Merge branch 'origin/feature/new-project-no-internet' of https://github.com/purduesigbots/pros-cli into origin/feature/new-project-no-internet

omegaStag commented:
> I'd rather just fix `resolve_templates` to always be OK if a remote depot is unavailable.
> 
> e.g. try block around:
> 
> https://github.com/purduesigbots/pros-cli/blob/1d9f6bae72b7e44ca4329d337b3921b7c41ae6df/pros/conductor/conductor.py#L119-L124

![image](https://user-images.githubusercontent.com/57331513/143691608-9a26b461-7782-4efa-b4e5-c5bd07ec51de.png)

When I try to have the try in resolve_templates it generates a bunch of error messages, seemingly because it doesn't like calling itself... but makes the project anyways? I have no idea

WillXuCodes commented:
I'm not seeing the try except block anywhere, have you pushed it yet? Also, what are you doing with the exception after catching it? It seems like it's just getting printed and moving on.

omegaStag committed:
Move try block to resolve_templates

omegaStag committed:
Use urllib.request to check for internet

This thing spits out duplicate warnings (2 for kernel and 2 for okapilib) dunno why

omegaStag committed:
Shorten warning

omegaStag committed:
Modify offline check

omegaStag committed:
Update conductor.py

omegaStag committed:
Update conductor.py

omegaStag commented:
With internet
![image](https://user-images.githubusercontent.com/57331513/147484747-ff5e786d-84ba-4b21-8831-e5a2db0a1bc7.png)

No internet
![image](https://user-images.githubusercontent.com/57331513/147484783-c14a1c91-f6c0-48e9-b7fe-fd002dfb4f7f.png)
There's not really an easy way to get rid of the duplicate warnings but I think that having them is better than not having any indication of no connection at all.

WillXuCodes reviewed (dismissed):
I'm approving this so we can get this shipped out the door a little quicker. The cosmetic issue of the warning appearing twice should be a seperate issue we open after the PR and somebody else could handle it as their entry level issue :P . Great work!
